### PR TITLE
Implement Graph Coloring using Backtracking

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
## Description

See in the implementation.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

**Comment by TruongNhanNguyen** [comment]
Can you help a look on this PR, @vil02?

File: src/backtracking/graph_coloring.rs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    adj_matrix: Vec<Vec<bool>>,
+    num_colors: usize,
+) -> Option<Vec<Vec<usize>>> {
+    let mut graph_coloring = GraphColoring::new(adj_matrix, num_colors);
+    graph_coloring.find_solutions()
+}
+
+/// A struct representing a graph coloring problem.
+struct GraphColoring {
+    adj_matrix: Vec<Vec<bool>>,
```
Comment (vil02) [inline src/backtracking/graph_coloring.rs:25]: Is there a check ensuring that the `adj_matrix` (please rename) is a square?

Furthermore, in this case I guess you really want to have a directed graph (i.e. symmetric adjacency matrix).

File: src/backtracking/graph_coloring.rs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    num_colors: usize,
```
Comment (vil02) [inline src/backtracking/graph_coloring.rs:26]: This is redundant.
```suggestion
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/backtracking/graph_coloring.rs (hunk 1 above)
Comment (TruongNhanNguyen) [reply src/backtracking/graph_coloring.rs:25]: I will add an assertion to check the adjacency matrix for squares. I also adjust the implementation to work with directed graphs.

File: src/backtracking/graph_coloring.rs (hunk 2 above)
Comment (TruongNhanNguyen) [reply src/backtracking/graph_coloring.rs:26]: I will remove it.

File: src/backtracking/graph_coloring.rs (hunk 1 above)
Comment (TruongNhanNguyen) [reply src/backtracking/graph_coloring.rs:25]: We can create a custom error type to handle the cases where the adjacency matrix is empty or not squared.

**Comment by TruongNhanNguyen** [comment]
Can you review it, @vil02?

File: src/backtracking/graph_coloring.rs [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+        ),
+        test_no_valid_coloring: (
+            vec![
+                vec![false, true, true],
+                vec![true, false, true],
+                vec![true, true, false],
+            ],
+            2,
+            Ok(None::<Vec<Vec<usize>>>)
+        ),
```
Comment (vil02) [inline src/backtracking/graph_coloring.rs:283]:
```suggestion
        ),
        test_more_colors_than_nodes: (
            vec![
                vec![true, true],
                vec![true, true],
            ],
            3,
            Ok(Some(vec![
                vec![1, 2],
                vec![1, 3],
                vec![2, 1],
                vec![2, 3],
                vec![3, 1],
                vec![3, 2],
            ]))
        ),
        test_no_coloring_with_zero_colors: (
            vec![
                vec![true],
            ],
            0,
            Ok(None::<Vec<Vec<usize>>>)
        ),
```

File: src/backtracking/graph_coloring.rs [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    ///
+    /// * `vertex` - The current vertex to be colored.
+    /// * `num_colors` - The number of colors available for coloring the graph.
+    fn find_colorings(&mut self, vertex: usize, num_colors: usize) {
+        if vertex == self.num_vertices() {
+            self.solutions.push(self.vertex_colors.clone());
+            return;
+        }
+
+        for color in 1..=num_colors {
```
Comment (vil02) [inline src/backtracking/graph_coloring.rs:111]: Is there a reason why you have chosen `1` to be the first color? It seams to be more natural (in languages like rust) to start with 0.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/backtracking/graph_coloring.rs (hunk 3 above)
Comment (TruongNhanNguyen) [reply src/backtracking/graph_coloring.rs:283]: It's a good idea.

File: src/backtracking/graph_coloring.rs (hunk 4 above)
Comment (TruongNhanNguyen) [reply src/backtracking/graph_coloring.rs:111]: There are no special reasons, I just use it on my own. After reviewing your suggestion, I see that it is more natural to start counting from 0, and using 0 as the first color can also make the implementation more consistent with typical array indexing.

**Comment by TruongNhanNguyen** [comment]
Let's have a final look, @vil02.

**Review by vil02 — APPROVED** [review]

**Pull request merged**